### PR TITLE
Cumulus 666 s3 credential endpoint

### PR DESCRIPTION
**Pull request by flamingbear**
**Summary:** Adds api endpoint `s3credentials` that allows an authorized EarthDataLogin user to retrieve temporary credentials for direct s3 access. 

Addresses [CUMULUS-666: Direct s3 access](https://bugs.earthdata.nasa.gov/browse/CUMULUS-666)

## Changes

* Added Endpoint at `s3credentials`
* Added unit and integration tests
* Updated travis to deploy s3AccessTest to us-west-2

## PR Checklist

- [ X ] Update CHANGELOG
- [ X ] Unit tests
- [ X ] Adhoc testing
- [ X ] Integration tests

### CODE REVIEW

File: example/README.md [hunk 1]
```diff
@@ -29,18 +29,18 @@ NOTE: For this to work you need your default credentials to be credentials for t
 
 ### Additional deployment steps
 
-An S3 Access lambda is needed in the us-west-1 region to run the tests. To initially create the lambda, run:
+An S3 Access lambda is needed in the us-west-2 region to run the integration tests. To initially create the lambda, run:
 
 ```
-aws lambda create-function --region us-west-1  --function-name <STACK>-S3AccessTest --zip-file fileb://app/build/cloudformation/<ZIP>-S3AccessTest.zip  --role arn:aws:iam::<AWS_ACCOUNT_ID>:role/<PREFIX>-lambda-processing  --handler index.handler --runtime nodejs6.10 --profile ngap-sandbox
+aws lambda create-function --region us-west-2  --function-name <STACK>-S3AccessTest --zip-file fileb://app/build/cloudformation/<ZIP>-S3AccessTest.zip  --role arn:aws:iam::<AWS_ACCOUNT_ID>:role/<PREFIX>-lambda-processing  --handler index.handler --runtime nodejs8.10 --profile ngap-sandbox
```
Comment (flamingbear) [inline example/README.md]: ngap explicitly denies access in us-west-1 so we need to use us-west-2

File: example/app/config.yml [hunk 2]
```diff
@@ -199,6 +203,30 @@ cumulus-nightly:
   stackName: nightly # used by the nightly cron job
   stackNameNoDash: Nightly
 
+cumulus-sit:
```
Comment (flamingbear) [inline example/app/config.yml]: Unrelated to changes, just puts SIT deployment up near the top.

File: example/iam/config.yml [hunk 3]
```diff
@@ -42,6 +45,31 @@ cumulus-from-pr:
 cumulus-nightly:
   prefix: nightly # used by the nightly cron job
 
+# Nightly SIT stack
```
Comment (flamingbear) [inline example/iam/config.yml]: same thing, moving around the config for clarity.

File: packages/api/tests/endpoints/test-s3credentials.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
+'use strict';
+'use strict';
```
Comment (ifestus) [inline packages/api/tests/endpoints/test-s3credentials.js]: Extra `use strict` here.

File: example/README.md (hunk 1 above)
Comment (laurenfrederick) [reply example/README.md]: 👍

File: packages/api/endpoints/s3credentials.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @param {Object} req - express request object
+ * @param {Object} res - express response object
+ * @returns {Object} the express response object with object containing
+ *                   tempoary s3 credentials for direct same-region s3 access.
+ */
+async function s3credentials(req, res) {
+  const username = req.authorizedMetadata.userName;
+  const credentials = await requestTemporaryCredentialsFromNgap(username);
+  return res.send(credentials.Payload);
```
Comment (laurenfrederick) [inline packages/api/endpoints/s3credentials.js]: The one thing here is when I tested the response came back as a JSON string and that is inconsistent with the rest of our API which comes back as JSON. I think maybe we should be sending back an object.

File: packages/api/endpoints/s3credentials.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * Use NGAP's time-based, temporary credential dispensing lambda.
+ *
+ * @param {string} username - earthdata login username
+ * @returns {Promise<Object>} Payload containing AWS STS credential object valid for 1
+ *                   hour.  The credential object contains keys: AccessKeyId,
+ *                   SecretAccessKey, SessionToken, Expiration and can be use
+ *                   for same-region s3 direct access.
+ */
+async function requestTemporaryCredentialsFromNgap(username) {
+  const FunctionName = 'gsfc-ngap-sh-s3-sts-get-keys';
```
Comment (markdboyd) [inline packages/api/endpoints/s3credentials.js]: Should this be an environment variable or something?

Comment (flamingbear) [reply packages/api/endpoints/s3credentials.js]: Good question, I had a comment above it before asking if we should be passing this in via ENVVAR.

I don't know the answer, I think it is probably wrapped up in what should this do for non-NGAP deploys.  NGAP has committed to supporting this lambda for users, I don't see that changing.  I suppose the only real value in making it a envvar would be to be able to update it on the lambda console without a redeploy.  It def made my eye twitch, but I don't know if adding it as an ENVVAR buys that much (it's particularly tied to this function)

Comment (laurenfrederick) [reply packages/api/endpoints/s3credentials.js]: I'm fine with it either way. I don't think it buys us a ton. This is a good point, though, and I think we need to call out in the API documentation (and any other documentation we end up writing) that this functionality is specific to NGAP.

Comment (flamingbear) [reply packages/api/endpoints/s3credentials.js]: @markdboyd talked with Lauren, I will update this with the option to use an environment variable and document that in non-ngap environments you can set that variable to a function that takes a username and returns time based security tokens via the AWS STS service.   But that it will default to the currently hard-coded function name for the NGAP dispenser.  

This seems like it will work fine, and if a non-ngap doesn't need credentials, you don't have to implement this functionality.  Seem reasonable?

Comment (flamingbear) [reply packages/api/endpoints/s3credentials.js]: a146eaa

File: example/spec/s3Access/s3AccessSpec.js [hunk 7]
```diff
@@ -90,36 +92,44 @@ describe('When accessing an S3 bucket directly', () => {
...
+    });
+
+    it('the expected user can assume same region access', async () => {
+      const credentials = new Credentials(creds);
+      const sts = new STS({ credentials });
+      const whoami = await sts.getCallerIdentity().promise();
+
+      expect(whoami.Arn).toMatch(/arn:aws:sts::\d{12}:assumed-role\/s3-same-region-access-role\/userName.*/);
+      expect(whoami.UserId).toMatch(/.*:userName.*/);
     });
```
Comment (flamingbear) [inline example/spec/s3Access/s3AccessSpec.js]: the getCallerIdentity on the token returns something like this 
```
    {
      "Arn": "arn:aws:sts::596205514787:assumed-role/s3-same-region-access-role/userNameb4dc82c935",
      "Account": "596205514787",
      "UserId": "AROAIH4VDHQ3M6PQUWHWM:userNameb4dc82c935",
      "ResponseMetadata": {
        "RequestId": "efa6cd7c-4023-11e9-8a35-b79d7bdd24ad"
      }
    }
```
I modified the test setup to use randomId instead of randomString, so I could match the userName that was created during the call to the api.  I didn't see a way to actually retrieve the userName directly to compare the whole thing.  But this seems good enough to me.

File: example/spec/s3Access/s3AccessSpec.js [hunk 8]
```diff
@@ -45,38 +81,76 @@ describe('When accessing a bucket directly', () => {
...
 
-    it('the data can be downloaded from within the same region', () => {
+    it('the expected user can assume same region access', async () => {
+      const credentials = new Credentials(creds);
+      const sts = new STS({ credentials });
+      const whoami = await sts.getCallerIdentity().promise();
 
+      expect(whoami.Arn).toMatch(/arn:aws:sts::\d{12}:assumed-role\/s3-same-region-access-role\/userName.*/);
+      expect(whoami.UserId).toMatch(/.*:userName.*/);
     });
```
Comment (flamingbear) [inline example/spec/s3Access/s3AccessSpec.js]: Also to be clear, only this one test is currently being run, the rest should work after NGAP work.
